src/util/proteomecluster.cpp

```diff
@@ -0,0 +1,492 @@
+    }
+    
+    unsigned int swMode = Alignment::initSWMode(par.alignmentMode, par.covThr, par.seqIdThr);
+
+    //Open the target protein database
+    DBReader<unsigned int> tProteinDB(par.db1.c_str(), par.db1Index.c_str(), par.threads, DBReader<unsigned int>::USE_DATA|DBReader<unsigned int>::USE_INDEX|DBReader<unsigned int>::USE_LOOKUP|DBReader<unsigned int>::USE_SOURCE);
+    tProteinDB.open(DBReader<unsigned int>::NOSORT);
+    const int tProteinSeqType = tProteinDB.getDbtype();
+
+    if (par.preloadMode != Parameters::PRELOAD_MODE_MMAP) {
+        tProteinDB.readMmapedDataInMemory();
+    }
+    //Open lookup table to get the source of the protein and the protein length
+    DBReader<unsigned int>::LookupEntry* tLookup = tProteinDB.getLookup();
+    const size_t tLookupSize = tProteinDB.getLookupSize();
+    unsigned int totalProteomeNumber = tLookup[tLookupSize - 1].fileNumber;
+    std::vector<ProteomeEntry> ProteomeList(totalProteomeNumber + 1);
+    calculateProteomeLength(ProteomeList, tLookup, tLookupSize, tProteinDB);
+
+    //Open the linclust result
+    DBReader<unsigned int> linResDB(par.db2.c_str(), par.db2Index.c_str(), par.threads, DBReader<unsigned int>::USE_DATA|DBReader<unsigned int>::USE_INDEX);
+    linResDB.open(DBReader<unsigned int>::LINEAR_ACCCESS);
+
+    //Open the DBWriter
+    DBWriter proteinClustWriter(par.db3.c_str(), par.db3Index.c_str(), par.threads, par.compressed, Parameters::DBTYPE_GENERIC_DB);
+    proteinClustWriter.open();
+    int proteomeDBType = DBReader<unsigned int>::setExtendedDbtype(Parameters::DBTYPE_GENERIC_DB, Parameters::DBTYPE_EXTENDED_SRC_IDENTIFIER);
+    DBWriter proteomeClustWriter(par.db4.c_str(), par.db4Index.c_str(), 1, par.compressed, proteomeDBType);
+    proteomeClustWriter.open();
+
+    std::vector<ClusterEntry> clusterReps; 
+
+    int gapOpen, gapExtend;
+    BaseMatrix *subMat;
+    subMat = new SubstitutionMatrix(par.scoringMatrixFile.values.aminoacid().c_str(), 2.0, par.scoreBias);
+    gapOpen = par.gapOpen.values.aminoacid();
+    gapExtend = par.gapExtend.values.aminoacid();
+    EvalueComputation evaluer(tProteinDB.getAminoAcidDBSize(), subMat, gapOpen, gapExtend);
+
+    // Debug(Debug::INFO) << "Start Initialization\n";
+    #pragma omp parallel
+    {
+        unsigned int thread_idx = 0;
+    #ifdef OPENMP
+        thread_idx = (unsigned int) omp_get_thread_num();
+    #endif    
+        std::vector<ClusterEntry> localClusterReps;
+        std::vector<unsigned int> localMemCount(ProteomeList.size(), 0);
+
+        #pragma omp for schedule(dynamic, 1)
+        for (size_t id = 0; id < linResDB.getSize(); id++) {
+            initLocalClusterReps(id, localClusterReps, localMemCount, tProteinDB, linResDB, thread_idx);
+        }
+
+        #pragma omp critical
+        {
+            for (size_t idx=0; idx < localMemCount.size(); idx++){
+                ProteomeList[idx].incrementMemCount(localMemCount[idx]);
```

Comment:
if you use __sync_fetch_and_add inside `incrementMemCount` you don't need the critical